homeassistant/components/tmb/sensor.py

```diff
@@ -0,0 +1,125 @@
+"""Support for TMB (Transports Metropolitans de Barcelona) Barcelona public transport."""
+from datetime import timedelta
+import logging
+
+from requests import HTTPError
+from tmb import IBus
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import ATTR_ATTRIBUTION, CONF_NAME
+from homeassistant.exceptions import Unauthorized
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTRIBUTION = "Data provided by Transport Metropolitans de Barcelona"
+
+ICON = "mdi:bus-clock"
+
+CONF_APP_ID = "app_id"
+CONF_APP_KEY = "app_key"
+CONF_LINE = "line"
+CONF_BUS_STOP = "stop"
+CONF_BUS_STOPS = "stops"
+ATTR_BUS_STOP = "stop"
+ATTR_LINE = "line"
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+
+LINE_STOP_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_BUS_STOP): cv.string,
+        vol.Required(CONF_LINE): cv.string,
+        vol.Optional(CONF_NAME): cv.string,
+    }
+)
+
+ROUTES_SCHEMA = vol.All(cv.ensure_list, [LINE_STOP_SCHEMA])
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_APP_ID): cv.string,
+        vol.Required(CONF_APP_KEY): cv.string,
+        vol.Optional(CONF_BUS_STOPS): ROUTES_SCHEMA,
```

Comment:
Since this is the only place `ROUTES_SCHEMA` is used, inline it:

```
PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_APP_ID): cv.string,
        vol.Required(CONF_APP_KEY): cv.string,
        vol.Optional(CONF_BUS_STOPS): vol.All(cv.ensure_list, [LINE_STOP_SCHEMA]),
    }
)
```

Reply:
Also, what happens if `CONF_BUS_STOPS`, which is marked as optional, isn't provided?

Reply:
If `CONF_BUS_STOPS` is not defined, it will not do anything. It could be useful if we don't want to track bus lines for a small period but we don't want to remove the integration (I know they are different ways to do it). Do you think that it's not worthy and we should make it required?

Reply:
I would make it required since that's where the intended functionality exists.

Reply:
fixed :)